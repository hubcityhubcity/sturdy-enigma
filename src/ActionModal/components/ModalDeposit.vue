<template>
  <ActionModal
    id="modal-deposit"
    ref="actionModal"
    :validate="validateForm"
    :amount="amount"
    title="Deposit"
    class="modal-deposit"
    submission-error-prefix="Depositing failed"
    :transaction-data="transactionData"
    :notify-message="notifyMessage"
    @close="clear"
    @txIncluded="onSuccess"
  >
    <TmFormGroup
      :error="$v.amount.$error && $v.amount.$invalid"
      class="action-modal-form-group"
      field-id="amount"
      field-label="Amount"
    >
      <span class="input-suffix">{{ denom | viewDenom }}</span>
      <TmField id="amount" v-model="amount" type="number" />
      <TmFormMsg
        v-if="balance.amount === 0"
        :msg="`doesn't have any ${viewDenom(denom)}s`"
        name="Wallet"
        type="custom"
      />
      <TmFormMsg
        v-else-if="$v.amount.$error && (!$v.amount.required || amount === 0)"
        name="Amount"
        type="required"
      />
      <TmFormMsg
        v-else-if="$v.amount.$error && !$v.amount.decimal"
        name="Amount"
        type="numeric"
      />
      <TmFormMsg
        v-else-if="$v.amount.$error && !$v.amount.between"
        :max="$v.amount.$params.between.max"
        :min="$v.amount.$params.between.min"
        name="Amount"
        type="between"
      />
    </TmFormGroup>
  </ActionModal>
</template>

<script>
import { mapGetters } from "vuex"
import gql from "graphql-tag"
import { uatoms, viewDenom, SMALLEST } from "src/scripts/num"
import { between, decimal } from "vuelidate/lib/validators"
import TmField from "src/components/common/TmField"
import TmFormGroup from "src/components/common/TmFormGroup"
import TmFormMsg from "src/components/common/TmFormMsg"
import ActionModal from "./ActionModal"
import transaction from "../utils/transactionTypes"
import { toMicroDenom } from "src/scripts/common"

export default {
  name: `modal-deposit`,
  components: {
    ActionModal,
    TmField,
    TmFormGroup,
    TmFormMsg
  },
  filters: {
    viewDenom
  },
  props: {
    proposalId: {
      type: [Number, String],
      required: true
    },
    proposalTitle: {
      type: String,
      required: true
    },
    denom: {
      type: String,
      required: true
    }
  },
  data: () => ({
    amount: 0,
    balance: {
      amount: 0,
      denom: ``
    }
  }),
  computed: {
    ...mapGetters([`network`]),
    ...mapGetters({ userAddress: `address` }),
    transactionData() {
      return {
        type: transaction.DEPOSIT,
        proposalId: this.proposalId,
        amounts: [
          {
            amount: uatoms(this.amount),
            denom: toMicroDenom(this.denom)
          }
        ]
      }
    },
    notifyMessage() {
      return {
        title: `Successful deposit!`,
        body: `You have successfully deposited your ${viewDenom(
          this.denom
        )}s on proposal #${this.proposalId}`
      }
    }
  },
  validations() {
    return {
      amount: {
        required: x => !!x && x !== `0`,
        decimal,
        between: between(SMALLEST, this.balance.amount)
      }
    }
  },
  methods: {
    viewDenom,
    open() {
      this.$refs.actionModal.open()
    },
    validateForm() {
      this.$v.$touch()

      return !this.$v.$invalid
    },
    clear() {
      this.$v.$reset()

      this.amount = 0
    },
    onSuccess(event) {
      this.$emit(`success`, event)
    }
  },
  apollo: {
    balance: {
      query: gql`
        query BalanceModalDeposit(
          $networkId: String!
          $address: String!
          $denom: String!
        ) {
          balance(networkId: $networkId, address: $address, denom: $denom) {
            amount
            denom
          }
        }
      `,
      skip() {
        return !this.userAddress
      },
      variables() {
        return {
          networkId: this.network,
          address: this.userAddress,
          denom: this.denom
        }
      }
    }
  }
}
</script>
