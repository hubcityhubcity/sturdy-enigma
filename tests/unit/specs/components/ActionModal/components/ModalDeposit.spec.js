"use strict"

import Vuelidate from "vuelidate"
import { shallowMount, createLocalVue } from "@vue/test-utils"
import ModalDeposit from "src/ActionModal/components/ModalDeposit"
import mockValues from "tests/unit/helpers/mockValues.js"

describe(`ModalDeposit`, () => {
  let wrapper, $store

  const localVue = createLocalVue()
  localVue.use(Vuelidate)
  localVue.directive("focus", () => {})

  const balance = { denom: `ATOM`, amount: `20` }

  beforeEach(async () => {
    $store = {
      commit: jest.fn(),
      dispatch: jest.fn(),
      getters: {
        userAddress: "cosmo1",
        network: "testnet"
      }
    }

    wrapper = shallowMount(ModalDeposit, {
      localVue,
      mocks: {
        $store
      },
      propsData: {
        proposalId: `1`,
        proposalTitle: mockValues.state.proposals[`1`].title,
        denom: `uatom`
      },
      sync: false
    })
    wrapper.setData({ balance })
  })

  it(`should display deposit modal form`, async () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it(`opens`, () => {
    const $refs = { actionModal: { open: jest.fn() } }
    ModalDeposit.methods.open.call({ $refs })
    expect($refs.actionModal.open).toHaveBeenCalled()
  })

  it(`clears on close`, () => {
    const self = {
      $v: { $reset: jest.fn() },
      amount: 10
    }

    ModalDeposit.methods.clear.call(self)
    expect(self.$v.$reset).toHaveBeenCalled()
    expect(self.amount).toBe(0)
  })

  it(`sends an event on success`, () => {
    const self = {
      $emit: jest.fn()
    }
    ModalDeposit.methods.onSuccess.call(self)
    expect(self.$emit).toHaveBeenCalledWith(
      "success",
      expect.objectContaining({})
    )
  })

  describe(`validation`, () => {
    describe(`fails`, () => {
      it(`with default values`, () => {
        expect(wrapper.vm.validateForm()).toBe(false)
      })

      it(`when the amount deposited higher than the user's balance`, async () => {
        wrapper.setData({ balance: { denom: `ATOM`, amount: `20` } })
        wrapper.setData({ amount: 250 })
        expect(wrapper.vm.validateForm()).toBe(false)
        await wrapper.vm.$nextTick()

        expect(wrapper.html()).toContain(`error="true"`)
      })

      it(`when the user doesn't have the deposited coin`, () => {
        wrapper.setData({ amount: 25 })
        expect(wrapper.vm.validateForm()).toBe(false)
      })
    })

    describe(`succeeds`, () => {
      it(`when the user has enough balance to submit a deposit`, async () => {
        wrapper.setData({ balance: { denom: `ATOM`, amount: `20` } })
        wrapper.setData({ amount: 10 })
        expect(wrapper.vm.validateForm()).toBe(true)
      })
    })
  })

  it("should return transaction data in correct form", () => {
    wrapper.setData({
      amount: 2
    })
    expect(wrapper.vm.transactionData).toEqual({
      type: "MsgDeposit",
      proposalId: "1",
      amounts: [
        {
          amount: "2000000",
          denom: "uatom"
        }
      ]
    })
  })

  it("should return notification message", () => {
    wrapper.setData({
      amount: 2
    })
    expect(wrapper.vm.notifyMessage).toEqual({
      title: `Successful deposit!`,
      body: `You have successfully deposited your ATOMs on proposal #1`
    })
  })
})
