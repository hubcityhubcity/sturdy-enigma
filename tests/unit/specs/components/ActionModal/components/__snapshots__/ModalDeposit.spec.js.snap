// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalDeposit should display deposit modal form 1`] = `
<actionmodal-stub
  amount="0"
  class="modal-deposit"
  id="modal-deposit"
  notifymessage="[object Object]"
  rewards=""
  submissionerrorprefix="Depositing failed"
  title="Deposit"
  transactiondata="[object Object]"
  validate="function () { [native code] }"
>
  <tmformgroup-stub
    class="action-modal-form-group"
    fieldid="amount"
    fieldlabel="Amount"
  >
    <span
      class="input-suffix"
    >
      ATOM
    </span>
     
    <tmfield-stub
      id="amount"
      type="number"
      value="0"
    />
     
    <!---->
  </tmformgroup-stub>
</actionmodal-stub>
`;
